### PR TITLE
fix(amf): fixed ambr calculations for different ambr related parameters.

### PR DESCRIPTION
**Pull request by Lkishor123**
Signed-off-by: LKishor123 <laawanya.kishor@wavelabs.ai>

## Summary
1. Fixed ambr calculations for different ambr related parameters.
2. Fixed sessiond sending same value for ul and dl ambr, even if different values are set in subscriber profile.

## Test Plan
1. Unit Testing
sit_3

2. Spirent

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    bit_rate_t* subscriber_ambr_ul) {
+  if (amf_ctxt_p == NULL) {
+    return RETURNerror;
+  }
+
+  *subscriber_ambr_dl = amf_ctxt_p->subscribed_ue_ambr.br_dl;
+  *subscriber_ambr_ul = amf_ctxt_p->subscribed_ue_ambr.br_ul;
+
+  return RETURNok;
+}
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp:36]: which units tests cover these newly added lines?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 2]
```diff
@@ -846,17 +846,18 @@ int amf_app_handle_pdu_session_accept(
...
+  smf_msg->msg.pdu_session_estab_accept.session_ambr.dl_unit =
+      smf_ctx->dl_ambr_unit;
+  smf_msg->msg.pdu_session_estab_accept.session_ambr.dl_session_ambr =
+      smf_ctx->dl_session_ambr;
+
+  smf_msg->msg.pdu_session_estab_accept.session_ambr.ul_unit =
+      smf_ctx->ul_ambr_unit;
+  smf_msg->msg.pdu_session_estab_accept.session_ambr.ul_session_ambr =
+      smf_ctx->ul_session_ambr;
+
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:860]: which unit test is covering these lined additions?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 3]
```diff
@@ -52,30 +53,11 @@ uint64_t get_bit_rate(uint8_t ambr_unit) {
...
-    if ((smf_context->ul_ambr_unit) < 4) {
-      *ul_pdu_ambr =
-          4 ^ (smf_context->ul_ambr_unit - 1) * (smf_context->ul_session_ambr);
-    } else {
-      *ul_pdu_ambr = smf_context->ul_session_ambr *
-                     get_bit_rate(smf_context->ul_ambr_unit);
-    }
-  }
+  *ul_pdu_ambr =
+      smf_context->ul_session_ambr * get_bit_rate(smf_context->ul_ambr_unit);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: which unit test is covering these?

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp (hunk 1 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp:36]: These line are getting the configured value for `ue aggregate max bit rate`. These are covered by end to end procedure unit test in: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp, TEST_F(AMFAppProcedureTest, TestPDUSessionSetup).

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 2 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:860]: These line are part of clean up and are covered by end to end procedure unit test in: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp, TEST_F(AMFAppProcedureTest, TestPDUSessionSetup).

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp (hunk 3 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: I'll add unit test for ambr_calculation_pdu_session().

**Comment by Lkishor123** [comment]
UERANSIM testing for basic registration and PDU session establishment accept:

sit_ambr_fix

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp (hunk 3 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: @ulaskozat, Unit tests are added, please review.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp (hunk 1 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_context.cpp:36]: These values(default_max_dl_bit_rate/default_max_ul_bit_rate) are configured in : lte/gateway/configs/subscriberdb.yml. This function `amf_smf_context_ue_aggregate_max_bit_rate_get()` is fetching it from amt_context which fetches it from subscriberdb.yml (using `amf_smf_context_ue_aggregate_max_bit_rate_set()`).

**Review by ulaskozat — APPROVED** [review]
LGTM

**Comment by ulaskozat** [comment]
@themarwhal or @uri200 can you stamp? looks like PR includes some minor changes on sessiond.

**Review by uri200 — APPROVED** [review]

**Pull request merged**